pennylane/ops/op_math/composite.py

```diff
@@ -129,8 +129,18 @@ def num_wires(self):
 
     @property
     def num_params(self):
+        """Compute the number of parameters as the sum of the numbers of parameters
+        of the constituents."""
         return sum(op.num_params for op in self)
 
+    @property
+    def grad_recipe(self):
+        """The gradient recipe of the composite operator. By defauly, the recipe is
+        set to None (no parameters) or ([None]*num_params) (parameters)."""
+        if (num_params := self.num_params) == 0:
+            return None
+        return [None] * num_params
```

Comment:
What about something like:
```
recipes = []
for op in self:
    if op.grad_recipe is not None:
        recipes.extend(op.grad_recipe)
return recipes
```

Or something.

For something like `qml.RX(x, 0) @ qml.RY(y, 0)`, we should be able to differentiate each component via parameter shift.

Reply:
Yes, for `Prod` we could have such a constructive method! It would be important to have the recipes map to the parameters of the `Prod` operator, though, and I'm not sure that's obviously guaranteed. E.g. if there is a single-parameter op with `grad_recipe=None` within the `Prod`, it's gonna shift the recipes<>parameters mapping.

Probably we could do
```python
recipes = sum(getattr(op, "grad_recipe", []) for op in self, start=[])
```
but I feel like this opens up a can of worms and is more than a bug fix...

Reply:
If `grad_recipe=None`, would it have any parameters?

Reply:
Yes, it could, it's the default value for `Operation`.